File: app/src/main/java/com/timotiusoktorio/pencake/ui/cart/CartViewModel.java
package com.timotiusoktorio.pencake.ui.cart;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.timotiusoktorio.pencake.data.model.CartItem;
import com.timotiusoktorio.pencake.data.model.ContactInfo;
import com.timotiusoktorio.pencake.data.model.Order;
import com.timotiusoktorio.pencake.data.model.OrderStatus;
import com.timotiusoktorio.pencake.data.model.State;
import com.timotiusoktorio.pencake.data.model.StoreInfo;
import com.timotiusoktorio.pencake.data.source.DataManager;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

class CartViewModel extends ViewModel {

    final MutableLiveData<State> stateLiveData = new MutableLiveData<>();
    final MutableLiveData<List<CartItem>> cartLiveData = new MutableLiveData<>();
    final MutableLiveData<Integer> quantityLiveData = new MutableLiveData<>();
    final MutableLiveData<Integer> subtotalLiveData = new MutableLiveData<>();
    final MutableLiveData<ContactInfo> contactInfoLiveData = new MutableLiveData<>();
    final MutableLiveData<Calendar> pickupScheduleLiveData = new MutableLiveData<>();
    final MutableLiveData<StoreInfo> storeInfoLiveData = new MutableLiveData<>();
    final MutableLiveData<Boolean> exitFlagLiveData = new MutableLiveData<>();

    private final DataManager dataManager;

    CartViewModel(DataManager dataManager) {
        this.dataManager = dataManager;
        loadCart();
        loadContactInfo();
        loadPickupSchedule();
        loadStoreInfo();
    }

    void loadCart() {
        stateLiveData.setValue(State.LOADING);
        dataManager.fetchCart(new DataManager.Callback<CartItem>() {
            @Override
            public void onSuccess(List<CartItem> data) {
                cartLiveData.setValue(data);
                calculateQuantityAndSubtotal(data);
                stateLiveData.setValue(State.SUCCESS);
            }

            @Override
            public void onError(String errorMsg) {
                stateLiveData.setValue(State.ERROR);
            }
        });
    }

    private void calculateQuantityAndSubtotal(List<CartItem> cart) {
        int quantity = 0;
        int subtotal = 0;
        for (CartItem cartItem : cart) {
            quantity += cartItem.getQuantity();
            subtotal += cartItem.getSubtotal();
        }
        quantityLiveData.setValue(quantity);
        subtotalLiveData.setValue(subtotal);
    }

    private void loadContactInfo() {
        ContactInfo contactInfo = dataManager.getOrderContactInfo();
        contactInfoLiveData.setValue(contactInfo);
    }

    private void loadPickupSchedule() {
        Date pickupScheduleDate = dataManager.getOrderPickupSchedule();
        if (pickupScheduleDate != null) {
            Calendar calendar = Calendar.getInstance();
            calendar.setTime(pickupScheduleDate);
            pickupScheduleLiveData.setValue(calendar);
        }
    }

    private void loadStoreInfo() {
        dataManager.fetchStoreInfo(new DataManager.SingleCallback<StoreInfo>() {
            @Override
            public void onSuccess(StoreInfo data) {
                storeInfoLiveData.setValue(data);
            }

            @Override
            public void onError(String errorMsg) {
            }
        });
    }

    void saveContactInfo(String name, String email, String phone) {
        ContactInfo contactInfo = new ContactInfo(name, email, phone);
        dataManager.saveOrderContactInfo(contactInfo);
    }

    void onDateSet(int year, int month, int dayOfMonth) {
        Calendar pickupSchedule = pickupScheduleLiveData.getValue();
        if (pickupSchedule == null) {
            pickupSchedule = Calendar.getInstance();
        }
        pickupSchedule.set(year, month, dayOfMonth);
        savePickupSchedule(pickupSchedule);
    }

    void onTimeSet(int hourOfDay, int minute) {
        Calendar pickupSchedule = pickupScheduleLiveData.getValue();
        if (pickupSchedule == null) {
            pickupSchedule = Calendar.getInstance();
        }
        int year = pickupSchedule.get(Calendar.YEAR);
        int month = pickupSchedule.get(Calendar.MONTH);
        int dayOfMonth = pickupSchedule.get(Calendar.DAY_OF_MONTH);
        pickupSchedule.set(year, month, dayOfMonth, hourOfDay, minute);
        savePickupSchedule(pickupSchedule);
    }

    private void savePickupSchedule(Calendar pickupSchedule) {
        pickupScheduleLiveData.setValue(pickupSchedule);
        dataManager.saveOrderPickupSchedule(pickupSchedule.getTime());
    }

    void clearCart() {
        dataManager.deleteCart();
        loadCart();
    }

    void placeOrder(ContactInfo contactInfo) {
        Long orderDate = new Date().getTime();
        List<CartItem> cart = cartLiveData.getValue();
        Integer orderSubtotal = subtotalLiveData.getValue();
        Long pickupSchedule = pickupScheduleLiveData.getValue() != null ? pickupScheduleLiveData.getValue().getTime().getTime() : 0;
        Order order = new Order(null, orderDate, cart, orderSubtotal, contactInfo, pickupSchedule, null, OrderStatus.ORDER_PLACED, false);
        dataManager.addOrder(order);
        exitFlagLiveData.setValue(true);
    }
}

File: app/src/main/java/com/timotiusoktorio/pencake/ui/cart/CartViewModel.kt
package com.timotiusoktorio.pencake.ui.cart

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.timotiusoktorio.pencake.data.model.*
import com.timotiusoktorio.pencake.data.source.DataManager
import java.util.*

class CartViewModel(private val dataManager: DataManager) : ViewModel() {

    val stateLiveData = MutableLiveData<State>()
    val cartLiveData = MutableLiveData<List<CartItem>>()
    val quantityLiveData = MutableLiveData<Int>()
    val subtotalLiveData = MutableLiveData<Int>()
    val contactInfoLiveData = MutableLiveData<ContactInfo>()
    val pickupScheduleLiveData = MutableLiveData<Calendar>()
    val storeInfoLiveData = MutableLiveData<StoreInfo>()
    val exitFlagLiveData = MutableLiveData<Boolean>()
    val placeOrderErrorFlagLiveData = MutableLiveData<Boolean>()

    init {
        loadCart()
        loadContactInfo()
        loadPickupSchedule()
        loadStoreInfo()
    }

    fun loadCart() {
        stateLiveData.value = State.LOADING
        dataManager.fetchCart(object : DataManager.Callback<CartItem> {
            override fun onSuccess(data: MutableList<CartItem>) {
                cartLiveData.value = data
                calculateQuantityAndSubtotal(data)
                stateLiveData.value = State.SUCCESS
            }

            override fun onError(errorMsg: String) {
                stateLiveData.value = State.ERROR
            }
        })
    }

    private fun calculateQuantityAndSubtotal(cart: List<CartItem>) {
        var quantity = 0
        var subtotal = 0
        for (cartItem in cart) {
            quantity += cartItem.quantity
            subtotal += cartItem.subtotal
        }
        quantityLiveData.value = quantity
        subtotalLiveData.value = subtotal
    }

    private fun loadContactInfo() {
        val contactInfo = dataManager.orderContactInfo
        contactInfoLiveData.value = contactInfo
    }

    private fun loadPickupSchedule() {
        dataManager.orderPickupSchedule?.let {
            val calendar = Calendar.getInstance()
            calendar.time = it
            pickupScheduleLiveData.value = calendar
        }
    }

    private fun loadStoreInfo() {
        dataManager.fetchStoreInfo(object : DataManager.SingleCallback<StoreInfo> {
            override fun onSuccess(data: StoreInfo) {
                storeInfoLiveData.value = data
            }

            override fun onError(errorMsg: String) {

            }
        })
    }

    fun saveContactInfo(name: String, email: String, phone: String) {
        dataManager.saveOrderContactInfo(ContactInfo(name, email, phone))
    }

    fun onDateSet(year: Int, month: Int, dayOfMonth: Int) {
        val pickupSchedule: Calendar = pickupScheduleLiveData.value ?: Calendar.getInstance()
        pickupSchedule.set(year, month, dayOfMonth)
        savePickupSchedule(pickupSchedule)
    }

    fun onTimeSet(hourOfDay: Int, minute: Int) {
        val pickupSchedule: Calendar = pickupScheduleLiveData.value ?: Calendar.getInstance()
        val year = pickupSchedule.get(Calendar.YEAR)
        val month = pickupSchedule.get(Calendar.MONTH)
        val dayOfMonth = pickupSchedule.get(Calendar.DAY_OF_MONTH)
        pickupSchedule.set(year, month, dayOfMonth, hourOfDay, minute)
        savePickupSchedule(pickupSchedule)
    }

    private fun savePickupSchedule(pickupSchedule: Calendar) {
        pickupScheduleLiveData.value = pickupSchedule
        dataManager.saveOrderPickupSchedule(pickupSchedule.time)
    }

    fun clearCart() {
        dataManager.deleteCart()
        loadCart()
    }

    fun placeOrder(name: String, email: String, phone: String, date: String, time: String) {
        if (name.isBlank() && email.isBlank() && phone.isBlank() && date.isBlank() && time.isBlank()) {
            placeOrderErrorFlagLiveData.value = true
            return
        }
        val order = Order(
                orderDate = Date().time,
                cart = cartLiveData.value ?: emptyList(),
                subtotal = subtotalLiveData.value ?: 0,
                contactInfo = ContactInfo(name, email, phone),
                pickupSchedule = pickupScheduleLiveData.value?.time?.time ?: 0
        )
        dataManager.addOrder(order)
        exitFlagLiveData.value = true
    }
}